rclcpp/src/rclcpp/callback_group.cpp

```diff
@@ -129,6 +138,31 @@ CallbackGroup::get_notify_guard_condition(const rclcpp::Context::SharedPtr conte
   return notify_guard_condition_;
 }
 
+rclcpp::GuardCondition::SharedPtr
+CallbackGroup::get_notify_guard_condition()
+{
+  std::lock_guard<std::recursive_mutex> lock(notify_guard_condition_mutex_);
+  auto context_ptr = this->get_context_();
+  if (context_ptr && context_ptr->is_valid()) {
+    std::lock_guard<std::recursive_mutex> lock(notify_guard_condition_mutex_);
+    if (notify_guard_condition_ && context_ptr != notify_guard_condition_->get_context()) {
+      if (associated_with_executor_) {
+        trigger_notify_guard_condition();
+      }
+      notify_guard_condition_ = nullptr;
+    }
+
+    if (!notify_guard_condition_) {
+      notify_guard_condition_ = std::make_shared<rclcpp::GuardCondition>(context_ptr);
+    }
+
+    return notify_guard_condition_;
+  } else {
+    throw std::runtime_error("Couldn't get guard condition from invalid context");
```

Comment:
Should we really throw an exception here?  That is, it sort of seem to me like failing to get the context might be a reasonable thing to happen, so should we just return nullptr and let the caller deal with it?  I'm not at all sure about this, so please take this as a question.

Reply:
Nope, the correct behavior should be to return nullptr, addressed in 87f41bf